crates/compilers/src/preprocessor/mod.rs

```diff
@@ -0,0 +1,257 @@
+
+/// Represents location of an item in the source map.
+/// Used to generate source code updates.
+#[derive(Debug)]
+pub struct SourceMapLocation {
+    /// Source map location start.
+    start: usize,
+    /// Source map location end.
+    end: usize,
+}
+
+impl SourceMapLocation {
+    /// Creates source map location from an item location within a source file.
+    fn from_span(source_map: &SourceMap, span: Span) -> Self {
+        let range = span.to_range();
+        let start_pos = BytePos::from_usize(range.start);
+        let end_pos = BytePos::from_usize(range.end);
+        Self {
+            start: source_map.lookup_byte_offset(start_pos).pos.to_usize(),
+            end: source_map.lookup_byte_offset(end_pos).pos.to_usize(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TestOptimizerPreprocessor;
+
+impl Preprocessor<SolcCompiler> for TestOptimizerPreprocessor {
+    fn preprocess(
+        &self,
+        _solc: &SolcCompiler,
+        mut input: SolcVersionedInput,
+        paths: &ProjectPathsConfig<SolcLanguage>,
+    ) -> Result<SolcVersionedInput> {
+        let sources = &mut input.input.sources;
+        // Skip if we are not preprocessing any tests or scripts. Avoids unnecessary AST parsing.
+        if sources.iter().all(|(path, _)| !is_test_or_script(path, paths)) {
+            trace!("no tests or sources to preprocess");
+            return Ok(input);
+        }
+
+        let sess = Session::builder().with_buffer_emitter(Default::default()).build();
+        let _ = sess.enter_parallel(|| -> solar_parse::interface::Result<()> {
+            let hir_arena = Arena::new();
+            let mut parsing_context = ParsingContext::new(&sess);
+            // Set remappings into HIR parsing context.
+            for remapping in &paths.remappings {
+                parsing_context
+                    .file_resolver
+                    .add_import_map(PathBuf::from(&remapping.name), PathBuf::from(&remapping.path));
+            }
+            // Load and parse test and script contracts only (dependencies are automatically
+            // resolved).
+            let preprocessed_paths = sources
+                .into_iter()
+                .filter(|(path, source)| {
+                    is_test_or_script(path, paths) && !source.content.is_empty()
+                })
+                .map(|(path, _)| path.clone())
+                .collect_vec();
+            parsing_context.load_files(&preprocessed_paths)?;
+
+            let hir = &parsing_context.parse_and_lower_to_hir(&hir_arena)?;
+            // Collect tests and scripts dependencies.
+            let deps = PreprocessorDependencies::new(
+                &sess,
+                hir,
+                &preprocessed_paths,
+                &paths.paths_relative().sources,
+            );
+            // Collect data of source contracts referenced in tests and scripts.
+            let data = collect_preprocessor_data(&sess, hir, deps.referenced_contracts.clone());
```

Comment:
clone not needed here

Reply:
yep, removed https://github.com/foundry-rs/compilers/pull/252/commits/f4a109d53bfe97bba287335404c2f5cc20e7dbf6